File: docs/g1/zh/python/galbot_robot/get_synced_observation.py
import time

from galbot_sdk.g1 import (
    GalbotRobot,
    JointState,
    JointStateMessage,
    RgbData,
    SensorType,
    SyncedObservation,
)

NS_TO_MS = 1e-6


def main():
    # 1) Initialize robot with sync mode enabled.
    #    The 2nd init arg must be True, otherwise get_synced_observation() is unavailable.
    robot = GalbotRobot()
    ok = robot.init(
        {
            SensorType.HEAD_LEFT_CAMERA,
            SensorType.LEFT_ARM_CAMERA,
            SensorType.RIGHT_ARM_CAMERA,
        },
        True,  # enable_sync_mode
    )
    if not ok:
        print("init failed")
        robot.destroy()
        return

    time.sleep(2)

    # 2) Build camera list for synchronization.
    #    cameras[0] is the anchor camera; other cameras are matched by nearest timestamp.
    cameras = [
        SensorType.LEFT_ARM_CAMERA,  # anchor
        SensorType.RIGHT_ARM_CAMERA,  # nearest-neighbor
        SensorType.HEAD_LEFT_CAMERA,  # nearest-neighbor
    ]

    # 3) Call API (typed return).
    obs: SyncedObservation | None = robot.get_synced_observation(cameras, True)
    if not obs:
        print("get_synced_observation failed")
        robot.request_shutdown()
        robot.wait_for_shutdown()
        robot.destroy()
        return

    # 4) Read synced RGB frames from typed field: obs.rgb_data_map
    print("[Synced RGB]")
    rgb_map: dict[SensorType, RgbData] = obs.rgb_data_map
    anchor_camera = cameras[0]
    anchor_frame: RgbData | None = rgb_map.get(anchor_camera, None)
    anchor_ts_ns = (
        anchor_frame.header.timestamp_ns if anchor_frame is not None else None
    )
    print(f"anchor_camera={anchor_camera}, anchor_timestamp_ns={anchor_ts_ns}")

    for cam in cameras:
        frame: RgbData | None = rgb_map.get(cam, None)
        cam_ts_ns = frame.header.timestamp_ns if frame is not None else None
        delta_ms_str = "N/A"
        if anchor_ts_ns is not None and cam_ts_ns is not None:
            delta_ms_str = f"{(cam_ts_ns - anchor_ts_ns) * NS_TO_MS:.3f}"
        print(
            f"camera={cam}, timestamp_ns={cam_ts_ns if cam_ts_ns is not None else 'N/A'}, "
            f"delta_to_anchor_ms={delta_ms_str}"
        )

    # 5) Read optional joint snapshot from typed field: obs.joint_state
    #    joint_state is JointStateMessage | None.
    joint_state: JointStateMessage | None = obs.joint_state
    print("[Joint]")
    joint_ts_ns = joint_state.timestamp_ns if joint_state is not None else None
    joint_delta_ms_str = "N/A"
    if anchor_ts_ns is not None and joint_ts_ns is not None:
        joint_delta_ms_str = f"{(joint_ts_ns - anchor_ts_ns) * NS_TO_MS:.3f}"
    print(
        "joint_timestamp_ns:",
        joint_ts_ns if joint_ts_ns is not None else "N/A",
        "delta_to_anchor_ms:",
        joint_delta_ms_str,
    )
    # 6) Each element in joint_state_vec is typed JointState and contains joint_name.
    joint_state_vec: list[JointState] = (
        joint_state.joint_state_vec if joint_state is not None else []
    )
    print("joint_count:", len(joint_state_vec))
    for i, js in enumerate(joint_state_vec[:5]):
        print(
            f"  [{i}] joint_name={js.joint_name}, "
            f"position={js.position}, "
            f"velocity={js.velocity}, "
            f"effort={js.effort}"
        )

    robot.request_shutdown()
    robot.wait_for_shutdown()
    robot.destroy()


if __name__ == "__main__":
    main()

File: examples/g1/cpp/galbot_robot/src/get_synced_observation.cpp
#include <algorithm>
#include <chrono>
#include <iostream>
#include <string>
#include <thread>
#include <unordered_set>
#include <vector>

#include "galbot_robot.hpp"

namespace {

std::string sensor_name(galbot::sdk::SensorType sensor) {
  using galbot::sdk::SensorType;
  switch (sensor) {
    case SensorType::HEAD_LEFT_CAMERA:
      return "HEAD_LEFT_CAMERA";
    case SensorType::HEAD_RIGHT_CAMERA:
      return "HEAD_RIGHT_CAMERA";
    case SensorType::LEFT_ARM_CAMERA:
      return "LEFT_ARM_CAMERA";
    case SensorType::RIGHT_ARM_CAMERA:
      return "RIGHT_ARM_CAMERA";
    case SensorType::LEFT_ARM_DEPTH_CAMERA:
      return "LEFT_ARM_DEPTH_CAMERA";
    case SensorType::RIGHT_ARM_DEPTH_CAMERA:
      return "RIGHT_ARM_DEPTH_CAMERA";
    default:
      return "UNKNOWN_CAMERA";
  }
}

}  // namespace

int main() {
  using namespace galbot::sdk;

  auto& robot = GalbotRobot::get_instance(MachineType::G1);
  std::unordered_set<SensorType> enable_sensor_set = {
      SensorType::HEAD_LEFT_CAMERA,
      SensorType::LEFT_ARM_CAMERA,
      SensorType::RIGHT_ARM_CAMERA,
  };
  if (!robot.init(enable_sensor_set, true)) {
    std::cout << "init failed" << std::endl;
    robot.destroy();
    return -1;
  }

  std::this_thread::sleep_for(std::chrono::milliseconds(300));

  const std::vector<SensorType> cameras = {
      SensorType::LEFT_ARM_CAMERA,   // anchor camera
      SensorType::RIGHT_ARM_CAMERA,  // nearest-neighbor aligned
      SensorType::HEAD_LEFT_CAMERA,  // nearest-neighbor aligned
  };
  auto obs = robot.get_synced_observation(cameras, true);
  if (!obs) {
    std::cout << "get_synced_observation failed" << std::endl;
    robot.request_shutdown();
    robot.wait_for_shutdown();
    robot.destroy();
    return -1;
  }

  std::cout << "[Synced RGB]" << std::endl;
  const auto anchor_it = obs->rgb_data_map.find(cameras[0]);
  if (anchor_it == obs->rgb_data_map.end() || !anchor_it->second) {
    std::cout << "  anchor missing" << std::endl;
  } else {
    const int64_t anchor_ts_ns = anchor_it->second->header.timestamp_ns;
    std::cout << "  anchor=" << sensor_name(cameras[0]) << " timestamp_ns=" << anchor_ts_ns << std::endl;
    for (const auto& cam : cameras) {
      const auto it = obs->rgb_data_map.find(cam);
      if (it == obs->rgb_data_map.end() || !it->second) {
        std::cout << "  " << sensor_name(cam) << ": missing" << std::endl;
        continue;
      }
      const int64_t cam_ts_ns = it->second->header.timestamp_ns;
      std::cout << "  " << sensor_name(cam) << " timestamp_ns=" << cam_ts_ns
                << " delta_to_anchor_ms=" << (cam_ts_ns - anchor_ts_ns) / 1e6
                << " bytes=" << it->second->data.size() << std::endl;
    }
  }

  if (obs->joint_state) {
    std::cout << "[Joint] timestamp_ns=" << obs->joint_state->timestamp_ns
              << " joint_count=" << obs->joint_state->joint_state_vec.size() << std::endl;
    const size_t print_n = std::min<size_t>(obs->joint_state->joint_state_vec.size(), 5);
    for (size_t i = 0; i < print_n; ++i) {
      const auto& js = obs->joint_state->joint_state_vec[i];
      std::cout << "  [" << i << "] joint_name=" << js.joint_name
                << " position=" << js.position
                << " velocity=" << js.velocity
                << " effort=" << js.effort << std::endl;
    }
  } else {
    std::cout << "[Joint] missing" << std::endl;
  }

  robot.request_shutdown();
  robot.wait_for_shutdown();
  robot.destroy();
  return 0;
}
